### PR TITLE
fix: Truncate framework generated query aliases.

### PR DESCRIPTION
**Pull request by SandPod**
## Change
- Truncates all framework generated query aliases to comply with [Postgres identifier limit of 63 bytes](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) (interpreted as 63 chars).

Closes: #1708 

## Truncation
Identifiers are truncated to 63 characters with the last 4 characters replaced by a hash created from the string of the full identifier. The identifier only needs to be unique in its own query context and is never stored.

The reason that only 4 characters are replaced is because that is what is used in the [Django ORM](https://github.com/django/django/blob/1b0a8991aec7079405be6b4be5fb17d69e1f4ab9/django/db/backends/utils.py#L281C5-L281C13) framework. And since its introduction it has not been changed so it seems safe to assume it is enough to mitigate collisions.

[Ruby on Rails](https://github.com/rails/rails/blob/90d0dc596133dc60839de121c5435712d84e21bf/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1563) uses 10 characters to store the hash.

## Guidelines for review
Best reviewed commit by commit to see the change with the related tests.

Currently we are missing proper unit tests for the changes in the TableRelation class. Since we have the integration tests that validate the functionality I think this is ready to be merged for the release on Monday. We will follow up with proper unit tests for the changes in the table relation class and its influence on the database query builder.

Issue for adding the unit tests: https://github.com/serverpod/serverpod/issues/1746
## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None, these are all framework generated.

### CODE REVIEW

File: packages/serverpod/lib/src/database/table_relation.dart [hunk 1]
```diff
@@ -61,12 +83,13 @@ class TableRelation {
 
   /// The query alias for field name to be joined on including all joins.
   String get fieldQueryAliasWithJoins {
-    return '${_fromRelationQueryAlias()}.${_tableRelationEntries.last.field.columnName}';
+    return _truncateFunction(
+        '${_fromRelationQueryAlias()}.${_tableRelationEntries.last.field.columnName}');
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/table_relation.dart:87]: NIT: Formatting looks strange here.

**Review by Isakdl — CHANGES_REQUESTED** [review]
Very nice to get this merged! 

A few comments and questions but nothing major.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/relation/relation_keyword_test.dart [hunk 2]
```diff
@@ -778,4 +779,189 @@ fields:
...
+    });
+
+    test('then implicit field is created', () {
+      var maybeField = classDefinition.fields
+          .where((element) => element.name != 'id')
+          .where((element) =>
+              element.name !=
+              'thisFieldIsExactly61CharactersLongAndIsThereforeAValidFieldNa')
+          .toList();
+      expect(maybeField, hasLength(1));
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/relation/relation_keyword_test.dart:816]: Will remove this specific check and only validate that we don't generate any fields that have a name longer than the Postgres name limit.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/relation/relation_keyword_test.dart [hunk 3]
```diff
@@ -778,4 +779,189 @@ fields:
...
+
+    test('then employee class has no field that is longer 63 chars.', () {
+      var fields = employeeDefinition?.fields
+          .where((element) => element.name.length > 63)
+          .map((e) => e.name)
+          .toList();
+
+      expect(fields, isEmpty,
+          reason: 'Expected no fields with a name longer than 63 chars');
+    }, skip: employeeDefinition == null);
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/relation/relation_keyword_test.dart:894]: Remove the explicit checks for the fields and only validate that we don't generate fields that exceed the Postgres name limit.

**Review by Isakdl — APPROVED** [review]
LGTM! 🚀

**Pull request merged**